pkg/services/secrets/kvstore/migrations/datasources_mig.go

```diff
@@ -0,0 +1,85 @@
+package migrations
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/grafana/grafana/pkg/bus"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/datasources"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/grafana/grafana/pkg/services/secrets/kvstore"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+const (
+	dataSourceSecretType = "datasource"
+)
+
+type DataSourceSecretMigrationService struct {
+	sqlStore           *sqlstore.SQLStore
+	dataSourcesService datasources.DataSourceService
+	secretsStore       kvstore.SecretsKVStore
+	features           featuremgmt.FeatureToggles
+	log                log.Logger
+	bus                bus.Bus
+}
+
+func ProvideDataSourceMigrationService(
+	cfg *setting.Cfg, dataSourcesService datasources.DataSourceService,
+	secretsStore kvstore.SecretsKVStore, features featuremgmt.FeatureToggles,
+	sqlStore *sqlstore.SQLStore, bus bus.Bus,
+) *DataSourceSecretMigrationService {
+	return &DataSourceSecretMigrationService{
+		sqlStore:           sqlStore,
+		dataSourcesService: dataSourcesService,
+		secretsStore:       secretsStore,
+		features:           features,
+		log:                logger,
+		bus:                bus,
+	}
+}
+
+func (s *DataSourceSecretMigrationService) Run(ctx context.Context) error {
+	return s.sqlStore.InTransaction(ctx, func(ctx context.Context) error {
+		query := &datasources.GetDataSourcesQuery{}
+		err := s.dataSourcesService.GetDataSources(ctx, query)
```

Comment:
I know I recommended moving the migration from the datasource service, but now when I look at this feels like the datasource secret migration is very much datasource service related and should live in there. In addition exposing these two new methods `DeleteDataSourceSecrets` and `DecryptLegacySecrets` on datasource service feels like a smell and you already have migration logic in the datasource service.

Moving back some things into datasource service would probably clean up the code a bit and at the same time keep similar logic in one place rather than two. WDYT?

Reply:
Sounds good, will move it back and make the methods private to the package.

It is very much datasource related, I think it makes sense to leave the migration on the service and the migrations provider on the secrets kvstore.

Reply:
Done on https://github.com/grafana/grafana/pull/50463/commits/0e16b5f23db9895bbd530243a56d3d056e7f34e5, https://github.com/grafana/grafana/pull/50463/commits/b04eb056b52c6ffe7ed90e7001c04c465a2d305f and https://github.com/grafana/grafana/pull/50463/commits/681862329d7699a67fe5f1a4189ec2e86db1a912.

The method `DeleteDataSourceSecrets` got removed on https://github.com/grafana/grafana/pull/50463/commits/bb1af49b905deceeff489b78f68247fd39e82977 in favor of https://github.com/grafana/grafana/pull/50463#discussion_r913141904.